[Maintenance] Add/update node & npm version checks

jasonblalock opened the pull request:
- Use package.json engines to validate node and npm versions via  check-node-version
- Set expected node version to 12.18.3
- Set expected npm version to ^6.14.6 to match match version shipped  with current node version 12.18.3
- As discussed in #4020. Originally related to #4003, however it was  decided not to use npm v7 at this time

jasonblalock committed:
[Maintenance] Add/update node & npm version checks

- Use package.json engines to validate node and npm versions via
  check-node-version
- Set expected node version to 12.18.3
- Set expected npm version to ^6.14.6 to match match version shipped
  with current node version 12.18.3
- As discussed in #4020. Originally related to #4003, however it was
  decided not to use npm v7 at this time

jasonblalock commented:
- Two different methods here, hopefully one is satisfactory
  - First is using package.json engines with the check-version npm
    script
    - Thinking we would completely replace the current preinstall script
      with this one
    - Validates everything based on the engines in package.json
    - Lose the ability for custom messaging, though the default
      messaging is decent
  - Second is updating the check-version.js script to also validate npm
    - A little clunkier as we have to exec out to get the npm version
      and that needs a callback. Open to suggestions for cleanup.
    - Hardcoded to just make sure npm >= 7

jasonblalock commented:
CI pipeline proving the scripts/check-version.js option is working

dimitropoulos commented on package.json:
```diff
@@ -10,6 +10,10 @@
   "homepage": "https://github.com/kong/insomnia#readme",
+  "engines": {
+    "node": "12.18.3",
+    "npm": ">=7"
```
I really want to update this (actually I want to just update to yarn... but that's another story), but I think there's a burden introduced by using an NPM version that didn't ship with this version of node.  Many people use [`n`](https://www.npmjs.com/package/n), for example, which syncs your NPM version as well:

```console
dimitri@dimitri:~$ n 16
  installing : node-v16.9.1
       mkdir : /home/dimitri/n/n/versions/node/16.9.1
       fetch : https://nodejs.org/dist/v16.9.1/node-v16.9.1-linux-x64.tar.xz
   installed : v16.9.1 to /home/dimitri/n/bin/node
      active : v16.9.1 at /usr/local/bin/node
dimitri@dimitri:~$ npm --version
7.21.1
dimitri@dimitri:~$ n 12.18.3
   installed : v12.18.3 to /home/dimitri/n/bin/node
      active : v12.18.3 at /usr/local/bin/node
dimitri@dimitri:~$ npm --version
6.14.6
```

It selects the version of `npm` that shipped with that version of node (https://nodejs.org/fr/blog/release/v12.18.3/).  I want to keep that parity there not just for development experience to say as simple as possible but also to avoid any unforeseen incompatibilities.

It's high on the team's mind to update node to a newer version, but that itself is tied to electron and we first need to update electron.  We did a fairly big electron recently and we are going to need to do another soon because the version we're on is going to be unsupported soon - so rest assured it's on our radar.

jasonblalock replied:
No problem. I only worked on this in response to @develohpanda's comments in #4003. If this isn't something your team is comfortable doing at the moment we can close both.

dimitropoulos replied:
Ahh, ok.  That makes more sense.  I need to read what the conversation was over on that other PR first.  My apologies if I said the opposite of @develohpanda - I happened to just see this first and had a moment on a Saturday to respond so thought I'd mention it while the thought was fresh.

One thing's for sure - any PR improving infrastructure or developer experience is highly cherished by ALL of us on the team!

develohpanda replied:
Dimitri raises a good point about using a version of NPM that's not shipped with the current version of Node, so perhaps we can pause on that momentarily. Using `check-node-version` instead of the script we have does seem like a _much_ nicer approach so I vote to still push this PR through and set the `engines.npm` version to be 6.14.6.

jasonblalock replied:
Alright I've updated the PR to use the check-node-version package (sorry about the force-push, wanted to fix the commits to have an accurate description). I've reset the package.json engines to the current node version and npm v6 (^6.14.6). I also updated the `scripts/check-version.js -> scripts/check-nvmrc-version.js` to instead make sure the .nvmrc matches the `engines.node`. Just to make sure those don't get out of sync whenever node is updated. It's naive in that it can't do any semver translation, but since it's a frozen node version I think it will suffice.

jasonblalock committed:
Re-purpose script to ensure .nvmrc matches package.json engines

- Rename `scripts/check-version.js` to `check-nvmrc-version` to match
  purpose
- Add `--yes` to npx command to ensure it runs without asking the user
  to install

jasonblalock committed:
Merge branch 'develop' into maintenance/add-npm-7-version-checks

jasonblalock committed:
Merge branch 'develop' into maintenance/add-npm-7-version-checks

develohpanda reviewed (approved):
LGTM!

Error when `.nvmrc` doesn't match `engines.node`
<img width="711" alt="image" src="https://user-images.githubusercontent.com/4312346/133361177-e6c8c628-8c2c-4e74-8d4b-f1cd46500488.png">

Error when `engines.node` doesn't match the installed version of node
<img width="581" alt="image" src="https://user-images.githubusercontent.com/4312346/133361247-77925b82-cf02-4ef9-999e-4c207d4dadad.png">

jasonblalock committed:
Merge branch 'develop' into maintenance/add-npm-7-version-checks

jackkav commented:
Closed as stale.
